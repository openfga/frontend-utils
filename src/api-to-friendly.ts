import { TypeDefinition, TypeDefinitions, Userset } from "@openfga/sdk";
import { Keywords } from "./keywords";

const readFrom = (obj: Userset, define: string[]) => {
  const childKeys = Object.keys(obj);

  childKeys.forEach((childKey: string) => {
    if (childKey === "this") {
      define.push(Keywords.SELF);
    }

    if (childKey === "tupleToUserset") {
      define.push(`${obj[childKey]!.computedUserset!.relation} ${Keywords.FROM} ${obj[childKey]!.tupleset!.relation}`);
    }

    if (childKey === "computedUserset") {
      define.push(`${obj[childKey]!.relation}`);
    }
  });
};

const apiToFriendlyRelation = (
  relation: string,
  relationDefinition: Userset,
  relations: string[],
  idx: number,
  newSyntax: string[],
) => {
  const relationKeys = Object.keys(relationDefinition);
  const define = [`    ${Keywords.DEFINE} ${relation}${relationKeys?.length ? ` ${Keywords.AS} ` : ""}`];

  // Read simple definitions
  readFrom(relationDefinition, define);

  relationKeys.forEach((relationKey) => {
    if (relationKey === "union") {
      const children = relationDefinition[relationKey]!.child!;
      children.forEach((child: Userset, idx: number) => {
        readFrom(child, define);

        if (idx < children.length - 1) {
          define.push(` ${Keywords.OR} `);
        }
      });
    }

    if (relationKey === "intersection") {
      const children = relationDefinition[relationKey]!.child!;
      children.forEach((child, idx: number) => {
        readFrom(child, define);

        if (idx < children.length - 1) {
          define.push(` ${Keywords.AND} `);
        }
      });
    }

    if (relationKey === "difference") {
      const { base, subtract } = relationDefinition[relationKey]!;

      readFrom(base!, define);
      define.push(` ${Keywords.BUT_NOT} `);
      readFrom(subtract!, define);
    }
  });

  // if (relations.length === idx + 1) {
  //   define.push("\n");
  // }

  newSyntax.push(define.join(""));
};

const apiToFriendlyType = (typeDef: TypeDefinition | TypeDefinition["relations"], newSyntax: string[]) => {
  if (typeDef.relations) {
    // A full type definition was passed
    newSyntax.push(`${Keywords.NAMESPACE} ${typeDef.type}`);
    if (Object.keys(typeDef.relations).length) {
      newSyntax.push(`  ${Keywords.RELATIONS}`);

      const relations = Object.keys(typeDef.relations);

      relations.forEach((relation: any, idx: number) => {
        const relationDefinition = (typeDef as TypeDefinition).relations[relation];
        apiToFriendlyRelation(relation, relationDefinition, relations, idx, newSyntax);
      });
    }
  } else {
    // A subset of the type definition with only the relations object was passed
    const relations = Object.keys(typeDef as TypeDefinition["relations"]);
    if (!relations.length) {
      return;
    }
    const relation = relations[0];
    const userSet = (typeDef as TypeDefinition["relations"])[relation];
    apiToFriendlyRelation(relation, userSet, relations, 0, newSyntax);
  }
};

export const apiSyntaxToFriendlySyntax = (
  config: TypeDefinitions | TypeDefinition,
  newSyntax: string[] = [],
): string => {
  const typeDefs = (config as TypeDefinitions)?.type_definitions;
  if (typeDefs) {
    typeDefs.forEach((typeDef) => {
      apiSyntaxToFriendlySyntax(typeDef, newSyntax);
    });
  } else if (config) {
    apiToFriendlyType(config as TypeDefinition, newSyntax);
  }

  return newSyntax.join("\n") + "\n";
};
