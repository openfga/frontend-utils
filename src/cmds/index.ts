/* eslint-disable import/namespace */
import { CommandModule } from "yargs";
import * as transform from "./transform";

export const commands: CommandModule[] = [transform as CommandModule];
