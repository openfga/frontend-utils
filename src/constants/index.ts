import { Keyword } from "./keyword";
import { SchemaVersion } from "./schema-version";
import { LANGUAGE_NAME } from "./language-name";

export const enums = {
  Keyword,
  SchemaVersion,
};

export { LANGUAGE_NAME };

const constants = {
  LANGUAGE_NAME,
  enums,
};

export default constants;
