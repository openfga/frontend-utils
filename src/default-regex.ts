// These rules should match the rules as specified in
// https://github.com/openfga/api/blob/main/openfga/v1/openfga.proto

// eslint-disable-next-line no-useless-escape
export const defaultTypeRule = "^[^:#@\\s]{1,254}$";
// eslint-disable-next-line no-useless-escape
export const defaultRelationRule = "^[^:#@\\s]{1,50}$";
