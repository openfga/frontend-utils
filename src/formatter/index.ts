import { indentDSL } from "./indent-dsl";

const formatter = {
  indentDSL,
};

export default formatter;
