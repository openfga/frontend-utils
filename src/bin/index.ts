#!/usr/bin/env node

import yargs from "yargs";
import { hideBin } from "yargs/helpers";
import { commands } from "../cmds";

yargs(hideBin(process.argv))
  .command(commands as any)
  .demandCommand()
  .recommendCommands()
  .strict()
  .help()
  .completion().argv;
