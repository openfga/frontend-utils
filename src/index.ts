#!/usr/bin/env node
import constants, { enums } from "./constants";
import formatter from "./formatter";
import transformer from "./transformer";
import validator from "./validator";
import * as syntaxHighlighters from "./syntax-highlighters";
import * as theming from "./theme";
import * as graphBuilder from "./utilities/graphs";
import sampleAuthorizationModels from "./samples";
// `language` imported like a module to avoid including antlr in the top level dependencies
// Local package path in package.json will be replaced with npm package version once published
import * as language from "./language/pkg/js";

const { Keyword, SchemaVersion } = enums;

const { friendlySyntaxToApiSyntax, apiSyntaxToFriendlySyntax } = transformer;
const checkDSL = validator.checkDSL;
const indentDSL = formatter.indentDSL;

export {
  sampleAuthorizationModels,
  constants,
  formatter,
  transformer,
  validator,
  syntaxHighlighters,
  theming,
  graphBuilder,
  language,

  // for backward compatibility to prevent breaking changes, to be removed in next major release
  Keyword,
  SchemaVersion,
  checkDSL,
  indentDSL,
  apiSyntaxToFriendlySyntax,
  friendlySyntaxToApiSyntax,
};
