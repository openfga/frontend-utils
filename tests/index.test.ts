import { apiSyntaxToFriendlySyntax, friendlySyntaxToApiSyntax } from "../src";

describe("friendlySyntaxToApiSyntax()", () => {
  describe("invalid syntax", () => {
    it("should throw if syntax is incorrect", () => {
      expect(() => {
        friendlySyntaxToApiSyntax(`model
  schema 1.0
t ype t1`);
      }).toThrowError();
    });

    it("should throw if syntax is incorrect", () => {
      expect(() => {
        friendlySyntaxToApiSyntax(`model
  schema 1.0
type test
  relations
    define writer as self and a but not test`);
      }).toThrowError();
    });

    it("should throw if syntax is incorrect", () => {
      expect(() => {
        friendlySyntaxToApiSyntax(`model
  schema 1.0
type test
  relations
    define writer as self and pepe or test `);
      }).toThrowError();
    });
  });

  it("should correctly read from `basic` definition", () => {
    const result = friendlySyntaxToApiSyntax(`model
  schema 1.0
type t1
  relations
    define r1 as self
`);

    const typeDf = result.type_definitions![0];
    const relations = typeDf.relations;

    expect(result.type_definitions).toHaveLength(1);
    expect(typeDf).toHaveProperty("type", "t1");
    expect(typeDf).toHaveProperty("relations");
    expect(relations).toHaveProperty("r1", { this: {} });
    expect(result).toMatchSnapshot();
  });

  it("should be able to handle single character type", () => {
    const result = friendlySyntaxToApiSyntax(`model
  schema 1.0
type t
  relations
    define r as self
`);

    const typeDf = result.type_definitions![0];
    const relations = typeDf.relations;

    expect(result.type_definitions).toHaveLength(1);
    expect(typeDf).toHaveProperty("type", "t");
    expect(typeDf).toHaveProperty("relations");
    expect(relations).toHaveProperty("r", { this: {} });
    expect(result).toMatchSnapshot();
  });

  it("should correctly read from `and` definition", () => {
    const result = friendlySyntaxToApiSyntax(`model
  schema 1.0
type t1
  relations
    define r1 as self and writer
`);

    const typeDf = result.type_definitions![0];
    const relations = typeDf.relations;
    const relation = relations!.r1;

    expect(result.type_definitions).toHaveLength(1);

    expect(typeDf).toHaveProperty("type", "t1");
    expect(typeDf).toHaveProperty("relations");

    expect(relations).toHaveProperty("r1");
    expect(relation).toHaveProperty("intersection");

    expect(relation.intersection).toHaveProperty("child");
    expect(relation.intersection!.child).toHaveLength(2);
    expect(relation.intersection!.child![0]).toEqual({ this: {} });
    expect(relation.intersection!.child![1]).toEqual({ computedUserset: { object: "", relation: "writer" } });
    expect(result).toMatchSnapshot();
  });

  it("should correctly read from `and` definition with $ as prefix", () => {
    const result = friendlySyntaxToApiSyntax(`model
  schema 1.0
type $t1
  relations
    define $r1 as self and $writer
`);

    const typeDf = result.type_definitions![0];
    const relations = typeDf.relations;
    const relation = relations!.$r1;

    expect(result.type_definitions).toHaveLength(1);

    expect(typeDf).toHaveProperty("type", "$t1");
    expect(typeDf).toHaveProperty("relations");
    expect(relations).toHaveProperty("$r1");
    expect(relation).toHaveProperty("intersection");

    expect(relation.intersection).toHaveProperty("child");
    expect(relation.intersection!.child).toHaveLength(2);
    expect(relation.intersection!.child![0]).toEqual({ this: {} });
    expect(relation.intersection!.child![1]).toEqual({ computedUserset: { object: "", relation: "$writer" } });
    expect(result).toMatchSnapshot();
  });

  it("should correctly read from `or` definition", () => {
    const result = friendlySyntaxToApiSyntax(`model
  schema 1.0
type t1
  relations
    define r1 as self or writer
`);

    const typeDf = result.type_definitions![0];
    const relations = typeDf.relations;
    const relation = relations!.r1;

    expect(result.type_definitions).toHaveLength(1);

    expect(typeDf).toHaveProperty("type", "t1");
    expect(typeDf).toHaveProperty("relations");

    expect(relations).toHaveProperty("r1");
    expect(relation).toHaveProperty("union");

    expect(relation.union).toHaveProperty("child");
    expect(relation.union!.child).toHaveLength(2);
    expect(relation.union!.child![0]).toEqual({ this: {} });
    expect(relation.union!.child![1]).toEqual({ computedUserset: { object: "", relation: "writer" } });
    expect(result).toMatchSnapshot();
  });

  it("should correctly read from `but not` definition", () => {
    const result = friendlySyntaxToApiSyntax(`model
  schema 1.0
type t1
  relations
    define r1 as self but not writer
`);
    const typeDef = result.type_definitions![0];
    const relations = typeDef.relations;
    const relation = relations!.r1;

    expect(result.type_definitions).toHaveLength(1);

    expect(typeDef).toHaveProperty("type", "t1");
    expect(typeDef).toHaveProperty("relations");

    expect(relations).toHaveProperty("r1");
    expect(relation).toHaveProperty("difference");

    expect(relation.difference!.base).toHaveProperty("this", {});
    expect(relation.difference!.subtract).toHaveProperty("computedUserset", { object: "", relation: "writer" });
    expect(result).toMatchSnapshot();
  });

  it("should correctly read from `from` definition", () => {
    const result = friendlySyntaxToApiSyntax(`model
  schema 1.0
type t1
  relations
    define share as owner from parent
`);

    expect(result.type_definitions).toHaveLength(1);
    expect(result).toMatchSnapshot();
  });

  it("should read a complex definition 1", () => {
    const result = friendlySyntaxToApiSyntax(`model
  schema 1.0
type folder
  relations
    define deleter as self
type doc
  relations
    define blocked_reader as self
    define delete as writer and deleter from parent
    define glass as writer
    define parent as self
    define read as reader or writer
    define reader as shared_reader but not blocked_reader
    define shared_reader as self
    define writer as self
    define creator as self
    define admin as self or creator
`);

    expect(result).toMatchSnapshot();
  });

  it("should read a complex definition 2", () => {
    const result = friendlySyntaxToApiSyntax(`model
  schema 1.0
type website
  relations
    define admin as self or owner
    define billing as self or admin
    define content_editor as self or admin
    define owner as self
    define viewer as self or admin or content_editor or billing
    define viewer as self and admin and content_editand and billing
`);

    expect(result).toMatchSnapshot();
  });

  it("should read a complex definition 3", () => {
    const result = friendlySyntaxToApiSyntax(`model
  schema 1.0
type website
  relations
    define admin as self or randy
    define billing as self or admin
    define content_editor as self or admin
    define randy as self
    define viewer as self or admin or content_editor or billing`);

    expect(result).toMatchSnapshot();
  });
});

describe("apiSyntaxToFriendlySyntax", () => {
  it("should correctly read from `basic` definition", () => {
    const result = apiSyntaxToFriendlySyntax({
      type_definitions: [
        {
          type: "t1",
          relations: {
            r1: {
              this: {},
            },
          },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  it("should correctly read from `and` definition", () => {
    const result = apiSyntaxToFriendlySyntax({
      type_definitions: [
        {
          type: "t1",
          relations: {
            r1: {
              intersection: {
                child: [
                  {
                    this: {},
                  },
                  {
                    computedUserset: {
                      object: "",
                      relation: "writer",
                    },
                  },
                ],
              },
            },
          },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  it("should correctly read from `or` definition", () => {
    const result = apiSyntaxToFriendlySyntax({
      type_definitions: [
        {
          type: "t1",
          relations: {
            r1: {
              union: {
                child: [
                  {
                    this: {},
                  },
                  {
                    computedUserset: {
                      object: "",
                      relation: "writer",
                    },
                  },
                ],
              },
            },
          },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  it("should correctly read from `but not` definition", () => {
    const result = apiSyntaxToFriendlySyntax({
      type_definitions: [
        {
          type: "t1",
          relations: {
            r1: {
              difference: {
                base: {
                  this: {},
                },
                subtract: {
                  computedUserset: {
                    object: "",
                    relation: "writer",
                  },
                },
              },
            },
          },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  it("should read a complex definition 4", () => {
    const result = apiSyntaxToFriendlySyntax({
      type_definitions: [
        {
          type: "folder",
          relations: {
            deleter: {
              this: {},
            },
          },
        },
        {
          type: "doc",
          relations: {
            blocked_reader: {
              this: {},
            },
            delete: {
              intersection: {
                child: [
                  {
                    computedUserset: {
                      object: "",
                      relation: "writer",
                    },
                  },
                  {
                    tupleToUserset: {
                      computedUserset: {
                        object: "",
                        relation: "deleter",
                      },
                      tupleset: {
                        object: "",
                        relation: "parent",
                      },
                    },
                  },
                ],
              },
            },
            glass: {
              computedUserset: {
                object: "",
                relation: "writer",
              },
            },
            parent: {
              this: {},
            },
            read: {
              union: {
                child: [
                  {
                    computedUserset: {
                      object: "",
                      relation: "reader",
                    },
                  },
                  {
                    computedUserset: {
                      object: "",
                      relation: "writer",
                    },
                  },
                ],
              },
            },
            reader: {
              difference: {
                base: {
                  computedUserset: {
                    object: "",
                    relation: "shared_reader",
                  },
                },
                subtract: {
                  computedUserset: {
                    object: "",
                    relation: "blocked_reader",
                  },
                },
              },
            },
            shared_reader: {
              this: {},
            },
            writer: {
              this: {},
            },
          },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });
});
