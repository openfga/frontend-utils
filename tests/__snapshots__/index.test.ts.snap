// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiSyntaxToFriendlySyntax should correctly read from \`and\` definition 1`] = `
"type t1
  relations
    define r1 as self and writer
"
`;

exports[`apiSyntaxToFriendlySyntax should correctly read from \`basic\` definition 1`] = `
"type t1
  relations
    define r1 as self
"
`;

exports[`apiSyntaxToFriendlySyntax should correctly read from \`but not\` definition 1`] = `
"type t1
  relations
    define r1 as self but not writer
"
`;

exports[`apiSyntaxToFriendlySyntax should correctly read from \`or\` definition 1`] = `
"type t1
  relations
    define r1 as self or writer
"
`;

exports[`apiSyntaxToFriendlySyntax should read a complex definition 4 1`] = `
"type folder
  relations
    define deleter as self
type doc
  relations
    define blocked_reader as self
    define delete as writer and deleter from parent
    define glass as writer
    define parent as self
    define read as reader or writer
    define reader as shared_reader but not blocked_reader
    define shared_reader as self
    define writer as self
"
`;

exports[`friendlySyntaxToApiSyntax() should be able to handle single character type 1`] = `
{
  "schema_version": "1.0",
  "type_definitions": [
    {
      "relations": {
        "r": {
          "this": {},
        },
      },
      "type": "t",
    },
  ],
}
`;

exports[`friendlySyntaxToApiSyntax() should correctly read from \`and\` definition 1`] = `
{
  "schema_version": "1.0",
  "type_definitions": [
    {
      "relations": {
        "r1": {
          "intersection": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "writer",
                },
              },
            ],
          },
        },
      },
      "type": "t1",
    },
  ],
}
`;

exports[`friendlySyntaxToApiSyntax() should correctly read from \`and\` definition with $ as prefix 1`] = `
{
  "schema_version": "1.0",
  "type_definitions": [
    {
      "relations": {
        "$r1": {
          "intersection": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "$writer",
                },
              },
            ],
          },
        },
      },
      "type": "$t1",
    },
  ],
}
`;

exports[`friendlySyntaxToApiSyntax() should correctly read from \`basic\` definition 1`] = `
{
  "schema_version": "1.0",
  "type_definitions": [
    {
      "relations": {
        "r1": {
          "this": {},
        },
      },
      "type": "t1",
    },
  ],
}
`;

exports[`friendlySyntaxToApiSyntax() should correctly read from \`but not\` definition 1`] = `
{
  "schema_version": "1.0",
  "type_definitions": [
    {
      "relations": {
        "r1": {
          "difference": {
            "base": {
              "this": {},
            },
            "subtract": {
              "computedUserset": {
                "object": "",
                "relation": "writer",
              },
            },
          },
        },
      },
      "type": "t1",
    },
  ],
}
`;

exports[`friendlySyntaxToApiSyntax() should correctly read from \`from\` definition 1`] = `
{
  "schema_version": "1.0",
  "type_definitions": [
    {
      "relations": {
        "share": {
          "tupleToUserset": {
            "computedUserset": {
              "object": "",
              "relation": "owner",
            },
            "tupleset": {
              "object": "",
              "relation": "parent",
            },
          },
        },
      },
      "type": "t1",
    },
  ],
}
`;

exports[`friendlySyntaxToApiSyntax() should correctly read from \`or\` definition 1`] = `
{
  "schema_version": "1.0",
  "type_definitions": [
    {
      "relations": {
        "r1": {
          "union": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "writer",
                },
              },
            ],
          },
        },
      },
      "type": "t1",
    },
  ],
}
`;

exports[`friendlySyntaxToApiSyntax() should read a complex definition 1 1`] = `
{
  "schema_version": "1.0",
  "type_definitions": [
    {
      "relations": {
        "deleter": {
          "this": {},
        },
      },
      "type": "folder",
    },
    {
      "relations": {
        "admin": {
          "union": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "creator",
                },
              },
            ],
          },
        },
        "blocked_reader": {
          "this": {},
        },
        "creator": {
          "this": {},
        },
        "delete": {
          "intersection": {
            "child": [
              {
                "computedUserset": {
                  "object": "",
                  "relation": "writer",
                },
              },
              {
                "tupleToUserset": {
                  "computedUserset": {
                    "object": "",
                    "relation": "deleter",
                  },
                  "tupleset": {
                    "object": "",
                    "relation": "parent",
                  },
                },
              },
            ],
          },
        },
        "glass": {
          "computedUserset": {
            "object": "",
            "relation": "writer",
          },
        },
        "parent": {
          "this": {},
        },
        "read": {
          "union": {
            "child": [
              {
                "computedUserset": {
                  "object": "",
                  "relation": "reader",
                },
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "writer",
                },
              },
            ],
          },
        },
        "reader": {
          "difference": {
            "base": {
              "computedUserset": {
                "object": "",
                "relation": "shared_reader",
              },
            },
            "subtract": {
              "computedUserset": {
                "object": "",
                "relation": "blocked_reader",
              },
            },
          },
        },
        "shared_reader": {
          "this": {},
        },
        "writer": {
          "this": {},
        },
      },
      "type": "doc",
    },
  ],
}
`;

exports[`friendlySyntaxToApiSyntax() should read a complex definition 2 1`] = `
{
  "schema_version": "1.0",
  "type_definitions": [
    {
      "relations": {
        "admin": {
          "union": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "owner",
                },
              },
            ],
          },
        },
        "billing": {
          "union": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "admin",
                },
              },
            ],
          },
        },
        "content_editor": {
          "union": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "admin",
                },
              },
            ],
          },
        },
        "owner": {
          "this": {},
        },
        "viewer": {
          "intersection": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "admin",
                },
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "content_editand",
                },
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "billing",
                },
              },
            ],
          },
        },
      },
      "type": "website",
    },
  ],
}
`;

exports[`friendlySyntaxToApiSyntax() should read a complex definition 3 1`] = `
{
  "schema_version": "1.0",
  "type_definitions": [
    {
      "relations": {
        "admin": {
          "union": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "randy",
                },
              },
            ],
          },
        },
        "billing": {
          "union": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "admin",
                },
              },
            ],
          },
        },
        "content_editor": {
          "union": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "admin",
                },
              },
            ],
          },
        },
        "randy": {
          "this": {},
        },
        "viewer": {
          "union": {
            "child": [
              {
                "this": {},
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "admin",
                },
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "content_editor",
                },
              },
              {
                "computedUserset": {
                  "object": "",
                  "relation": "billing",
                },
              },
            ],
          },
        },
      },
      "type": "website",
    },
  ],
}
`;
